### ISSUE TITLE
Support for pnpm

### ISSUE THREAD

**Issue by willsoto** [volta-cli/volta#737]
Seems to be a fairly popular alternative to `npm` and `yarn`. Apologies if this has already been suggested but I couldn't find an existing issue. It would be great to be able to `volta pin pnpm`.

https://pnpm.js.org/en/

**Comment by charlespierce** [comment]
Agreed, this is definitely something we’ve discussed on Discord, I’m surprised there wasn’t an issue already either. Thanks!

**Comment by arxpoetica** [comment]
What would need to take place for this to happen? Point me to the right place in the code, and I'll see if I can get around to a PR.

**Comment by charlespierce** [comment]
Hi @arxpoetica, we would definitely welcome that! Adding support for `pnpm`—or really any unsupported package manager—is a bit involved, though hopefully not too difficult. There are 3 main areas that need to be updated:

* Fetching / installing / pinning the version, which is all contained within the `tool` module in `crates/volta-core` (https://github.com/volta-cli/volta/tree/main/crates/volta-core/src/tool)
    * The changes here can mostly be based on what is done for `npm`, the only caveat I'm aware of is that instead of editing the launcher files, we need to create them (so that the bare command `pnpm` works as expected).
* Reading / writing `pnpm` settings, contained in the `platform`, `project`, and `toolchain` modules.
    * The changes here should mostly be adding the `pnpm` key alongside `npm` and `yarn`
* Running `pnpm` through the shim, contained in the `run` module.
    * Here the logic should be almost identical to the `yarn` logic.

That's a very high-level overview of the needed changes, but I'll take a look at write some more this week about any corner cases / difficulties I can predict. I'm also happy to pair on this for a while if that would help.

**Comment by arxpoetica** [comment]
@charlespierce this ended up being something I just haven't had time to get around to yet. It's definitely still on my radar, but if someone feels like giving it a head start (or even attacking totally)...

...I think you'll find it happening more quickly. Sorry for my lack of involvement (re: time).

**Comment by charlespierce** [comment]
@arxpoetica No worries at all, thanks for updating!

**Comment by willsoto** [comment]
@charlespierce slightly related to this, has any thought been given to pinning arbitrary dependencies that would useful to have available on the command-line? For example, it might be useful to have `mocha` or `tsc` installed and available directly on the command line at a specific version without telling users to `volta install` it themselves. Another example is something like `lerna` or `rush` which are the primary entry-points for repos that are managed by those tools.

Under the hood, maybe something specific happens when `pnpm` is installed but in some ways it is just another command line tool being managed by Volta.

Just wondering if you consider this out of scope since the [documentation for `pin`](https://docs.volta.sh/reference/pin) has something specifically about this.

**Comment by charlespierce** [comment]
Hi @willsoto, are you talking about something like #861? In that issue there's some discussion about the complications, however that is definitely something we'd like to support. Or do you mean a setting in `package.json` that says "Use Volta to manage running `lerna`"?

If the latter, there's a couple reasons that's not really in scope for Volta:

1. Technically speaking, Volta operates with shims in a shell-agnostic way, there's no way for us to shim tools that we don't know about ahead of time (if a user runs `lerna` and the shell says "Command not found", that never goes through anything Volta controls). Individual shells may have a way to control that behavior, but it's unlikely to be universally applicable across OS and shell combinations.
2. Philosophically, we decided against creating automatic shims for everything that had been installed, because it caused issues and raised some security concerns (see #235 for the discussions). We ultimately decided that the end user should have direct control over which tools we do or do not manage, so it would go against that to allow packages to take that control away from the user.

**Comment by willsoto** [comment]
Of course I didn't see #861 when I went looking for existing issues 🤦 

After reading through it a couple of times, it does _seem_ like that would also solve my problem, just in a different way.

> Or do you mean a setting in package.json that says "Use Volta to manage running lerna"?

Yeah, something like this is what I had in mind:

```json5
// package.json
{
  "volta": {
    "lerna": "2.0.0"
  }
}
```

I totally understand if this is considered out of scope for Volta, but I thought since `pnpm` is something that gets installed via `npm` (I don't think they have an alternative way to download?) then you could really get any tool in this way. I'll subscribe to #861 and see where that goes so I don't clutter this issue up. Thanks

**Comment by zkochan** [comment]
@muuvmuuv [suggested us](https://github.com/pnpm/pnpm/issues/3146) to document Volta as the suggested tool for pnpm. Unlike Yarn, pnpm currently doesn't have the ability to seamlessly switch between different versions of pnpm. So if Volta would support pnpm, I think we would just recommend using the pin feature of Volta.

Also, @arcanis might be interested in this discussion, he created [corepack](https://github.com/nodejs/corepack)

**Comment by zkochan** [comment]
I think we can offer a $50 bounty from our opencollective funds (if this is enough, I don't know how much work it involves).

**Comment by charlespierce** [comment]
@zkochan I took a look at it before our 1.0 release late last year, and it actually was somewhat involved, as our models and abstractions around the package managers aren't as clean as they could be (some room for tech debt cleanup). However, it was mostly straightforward. The main sticking point was around the global package directory: As I recall, there isn't a clean way to redirect _both_ the installation and `bin` directories to a custom location in pnpm using only environment variables.

Volta does that redirection when installing global packages so that we can sandbox them to a specific Node version, and I recall struggling to get that working. I _think_ I could do it with a command-line flag, but that gets a little tricky if there's already one specified and so I didn't dig in too deeply. There also appeared to be a "layout version" type directory that was always appended, which we would need to take into account, but I think I had worked out how we can handle that on our side.

I determined most of that with some experimentation and poking around the source of pnpm, so if there's something I missed about how to define the global directories, that would be much appreciated!

So in all, I think it's probably a larger task than a $50 bounty, however if someone was interested and wanted to collect that on top of the experience, I'd be more than happy to mentor them through it. Working on Volta is also part of my day-to-day, so if I can get it onto the roadmap that would be the other way.

**Comment by zkochan** [comment]
Is there something we could implement on pnpm's side to simplify the task?

You may set the `NPM_CONFIG_GLOBAL_DIR` env variable to change the global directory used by pnpm. The global bin directory will currently be selected automatically from one of the directories in PATH. pnpm has to have write access to the directory and the directory should have npm, pnpm, nodejs, or node in the path

https://github.com/pnpm/pnpm/blob/9c41e11b0bfa3f4ece2b7d4d67690496f1c9b9c0/packages/global-bin-dir/src/index.ts#L35-L40

**Comment by charlespierce** [comment]
> Is there something we could implement on pnpm's side to simplify the task?

I don't think so, only because those changes would necessarily be in only newer versions of pnpm, while we would ideally want to support a larger range of versions.

> You may set the `NPM_CONFIG_GLOBAL_DIR` env variable to change the global directory used by pnpm. The global bin directory will currently be selected automatically from one of the directories in PATH. pnpm has to have write access to the directory and the directory should have npm, pnpm, nodejs, or node in the path
> 
> https://github.com/pnpm/pnpm/blob/9c41e11b0bfa3f4ece2b7d4d67690496f1c9b9c0/packages/global-bin-dir/src/index.ts#L35-L40

Interesting, that is jogging my memory a bit. I think I may have been able to get that to work by modifying the PATH, but I don't recall the specific issues I was hitting any more (and I appear to have failed to write them down anywhere). I don't think any of them were insurmountable, just that they were decent amount of work to fit into everything, taking into account our current model.

**Comment by muuvmuuv** [comment]
I would add another 30€. 

Just saying, but I'm already using pnpm with Volta since three months without any problems but would also love to see a "clean" way of implementation for future package managers/binaries so Volta will stay on top of the other managers.

**Comment by cjk** [comment]
I'm using pnpm a lot, so I'll wait until proper pnpm-support is available before switching to Volta

**Comment by Kinrany** [comment]
> I don't think so, only because those changes would necessarily be in only newer versions of pnpm, while we would ideally want to support a larger range of versions.

Is it possible to provide support for newer versions and just warn the user that older versions are not supported?

I think most people using pnpm are OK with upgrading once even if they want to pin their version and not upgrade regularly.

**Comment by shiftgeist** [comment]
> I think most people using pnpm are OK with upgrading once even if they want to pin their version and not upgrade regularly.

`pnpm` warns about outdated versions whenever you use it. This should reduce the amount of outdated versions. At least for desktop use.

**Comment by charlespierce** [comment]
@Kinrany @shiftgeist Those are good points, if we need to we can likely work to get changes upstreamed to `pnpm` and then only support that version or later. Ideally we'd make it work without that, but as a last resort that makes sense, especially if pnpm itself is vocal about pushing you to upgrade.

**Comment by J141** [comment]
As I would also like to see pnpm officially supported as a pinnable package manager by Volta, I will be adding a $110 bounty for getting this supported. Together with zkochan's $50 bounty offer, this brings the total collectable bounty to $160. (+30EUR/35 USD if muuvmuuv is actually offering another 30EUR, but not sure whether he's offering €30 or was suggesting to raise the bounty by €30).

Hopefully this is enough to stimulate integration of pnpm :)

**Comment by arxpoetica** [comment]
I'll add $50 to that. This is one of my most watched issues.

**Comment by muuvmuuv** [comment]
Imagine they just pokering to get more people putting $$$ into the issue 😆

**Comment by J141** [comment]
> Imagine they just pokering to get more people putting $$$ into the issue 😆

The goal is to entice someone who's knowledgable enough to pick up this issue and integrate pnpm. The collectable bounty for implementing this is now $210. Surely someone must be interested in making a quick buck?

**Comment by muuvmuuv** [comment]
This was meant as a joke...

**Comment by shiftgeist** [comment]
Is there an issuehunt or similar project?

**Comment by ekilzen** [comment]
> Is there an issuehunt or similar project?

I don't think so...can't find the link.

**Comment by shiftgeist** [comment]
> > Is there an issuehunt or similar project?
> 
> I don't think so...can't find the link.

Might be useful for collecting the funds (doesn't need to be issuehunt tho). I imagine multiple paypal transaction being a little sketchy.

**Comment by charlespierce** [comment]
Hey everyone! Sorry about the delayed response, I've been out for a week or so. Super exciting to see so much interest for pnpm, I'd definitely be happy to help facilitate the bounty if someone wants to take it on (and help guide people through the code for that section). My time has been pretty limited lately, as I just started a new job, but I'll also try to carve some out to push this forward.

**Comment by mikrostew** [comment]
I have an RFC up for adding pnpm support: https://github.com/volta-cli/rfcs/pull/46

I think I captured most of the details and requirements from this and other conversations, and welcome your feedback!

**Comment by zkochan** [comment]
I have lost interest in this feature.

Node.js shipped Corepack, which autoinstalls the right version of pnpm: https://nodejs.org/dist/latest-v16.x/docs/api/corepack.html

Also, pnpm may be used as a Node.js version manager: https://pnpm.io/cli/env

**Comment by arxpoetica** [comment]
Corepack is still experimental.

**Comment by canadaduane** [comment]
pnpm (installed via volta) suggested I upgrade today: `Update available! 6.18.0 → 6.21.0.` and so I did what they suggested, "Run `pnpm add -g pnpm` to update" and pnpm is now broken:

```
$ pnpm install
Volta error: 'volta-shim' should not be called directly.

Please use the existing shims provided by Volta (node, yarn, etc.) to run tools.
```

Just running the executable does the same:
```
$ pnpm
Volta error: 'volta-shim' should not be called directly.

Please use the existing shims provided by Volta (node, yarn, etc.) to run tools.
```

I don't know how to get back to a working state, so I might need to remove volta.

**Comment by chriskrycho** [comment]
A workaround for that last bit while we're working on pnpm support: in a repo which needs pnpm, start by running `npm install` and *then* run `pnpm install`. After that, you'll be in a working state. (It's quite annoying, we all recognize! Hopefully we'll get some motion on merging and implementing volta-cli/rfcs#46 soon!)

**Comment by sbonasu** [comment]
can you please let me know if there are any updates on this feature request.

**Comment by charlespierce** [comment]
@sbonasu There aren't any updates at the moment. All of the maintainers are busy with other work. However, in the RFC linked above, there's a potential way forward after some feedback from the maintainer of `pnpm`, where we can likely start with supporting pnpm 7, which should be more straightforward.

When there are more updates, you should see them here.

**Comment by coconilu** [comment]
couldn't agree more, please let me know if there are any updates

**Comment by Mr-xzq** [comment]
Now that pnpm is very popular, expect volta to add support for pnpm.

**Comment by chawyehsu** [comment]
Hi all participants, I just initialized PR #1273 which is going to take this feature request into reality. I don't know if the bounty is still available but please consider leaving comments when you are available so I can push it forward.

**Comment by arxpoetica** [comment]
> I don't know if the bounty is still available

I'm still good for $50.

**Comment by st-jordy-de-koning** [comment]
I'm also still offering a $110 bounty.

**Comment by J141** [comment]
I'm also still offering a $110 bounty. It's yours if PR 1273 gets merged.

**Comment by IT-MikeS** [comment]
I will offer up $50 as well if pnpm support is added from your PR @chawyehsu

**Comment by ScreamZ** [comment]
What about this ? https://pnpm.io/cli/env Seems this is a war between both tools 😅 I prefer volta way to manage things btw

**Comment by NatalePorto** [comment]
Looking forward to see this issue closed and integrated in a new release.

**Comment by ild0tt0re** [comment]
please guys, the world is switching to pnpm let us continue to use our favorite tool manager volta ⚡ ❤️

**Comment by wagenet** [comment]
Is this resolved by #1273 ?

**Comment by IT-MikeS** [comment]
> Is this resolved by #1273 ?

Seems like it. As promised @chawyehsu I sent in a one-time sponsor of 50$. Thank you for your work!

**Comment by chawyehsu** [comment]
@wagenet The PR was successfully merged, the feature does not release yet though [CI builds are available](https://github.com/volta-cli/volta/actions/runs/3383641697) for a try. At present, there is no ETA for it, the team is likely to gather more fixes such as Linux ARM builds and the TLS cert fix apart from this feature for the next release.

@IT-MikeS Much appreciate!

**Comment by charlespierce** [comment]
I just opened #1394 to add a feature flag environment variable to ensure we maintain backwards compatibility for users who currently have `pnpm` as a global package. Once that is approved / merged, we should be able to release Volta 1.1.1 with pnpm support included. The feature flag will mean that anyone who wants to try this behavior will need to set `VOLTA_FEATURE_PNPM` to `1` in their environment variables or profile script, for example:

```shell
export VOLTA_HOME="$HOME/.volta"
export VOLTA_FEATURE_PNPM=1
export PATH="$VOLTA_HOME/bin:$PATH"
```

**Comment by haosmos** [comment]
> I just opened #1394 to add a feature flag environment variable to ensure we maintain backwards compatibility for users who currently have `pnpm` as a global package. Once that is approved / merged, we should be able to release Volta 1.1.1 with pnpm support included. The feature flag will mean that anyone who wants to try this behavior will need to set `VOLTA_FEATURE_PNPM` to `1` in their environment variables or profile script, for example:
> 
> ```shell
> export VOLTA_HOME="$HOME/.volta"
> export VOLTA_FEATURE_PNPM=1
> export PATH="$VOLTA_HOME/bin:$PATH"
> ```

@charlespierce, hi! 😃

Any news? When should we expect the long-awaited release with pnpm support? Is there any chance it will be released this year?

**Comment by charlespierce** [comment]
Sorry for the delay on this, I've been swamped with other work! Volta 1.1.1 was just released and it includes the experimental `pnpm` support (_HUGE_ thanks to @chawyehsu for the implementation)! I also added a section to the advanced docs about the [pnpm Support](https://docs.volta.sh/advanced/pnpm), noting the known limitations and how to activate it!

**Comment by arxpoetica** [comment]
@chawyehsu sent my sponsorship!

**Comment by pegak** [comment]
Great, thanks! I am already trying it.

I would also add to limitations that it is not possible to pin `pnpm`. :)

**Comment by chawyehsu** [comment]
@pegak pnpm pinning is supported in 1.1.1. Perhaps you are encountering some issues with it. Please feel free to open a new issue to let us know.

**Comment by pegak** [comment]
Oh, didn't have the update, sorry. Trying now!

**Comment by j** [comment]
@chawyehsu

```
~/Code/example main* ❯ volta --version
1.1.1

~/Code/example main* ❯ pnpm --version
7.27.1

~/Code/example main* ❯ volta pin pnpm@7.27.1
error: Only node and yarn can be pinned in a project

Use `npm install` or `yarn add` to select a version of pnpm for this project.
```

**Comment by j** [comment]
Also, pnpm run commands use global node:

```
 WARN  Unsupported engine: wanted: {"node":">=19.0.0"} (current: {"node":"v18.13.0","pnpm":"7.27.1"})
```

**Comment by j** [comment]
I just came across that doc.  Sort of a pain as things seem so disconnected when using pnpm.  pnpm is great and recommended by turborepo, but pnpm commands use global node, etc, etc.  Fun!

**Comment by chawyehsu** [comment]
Sorry for the inconvenience @j

> Also, pnpm run commands use global node:
> 
> ```
>  WARN  Unsupported engine: wanted: {"node":">=19.0.0"} (current: {"node":"v18.13.0","pnpm":"7.27.1"})
> ```

> but pnpm commands use global node, etc, etc

This warning is from pnpm and not relevant to the current implementation of pnpm support for Volta. In the implementation, we only intercept a basic set of pnpm commands to work with Volta in pinning, etc. And most of the workflows are directly passed to pnpm transparently. Hence there may be things unlinked when using pnpm at present because pnpm may work outside of Volta's control.

Talking about the warning message, that's because you specified the node version via `engines` in package.json for your project. Something like this:

```
{
  "engines": {
    "node": ">=19"
  }
}
```

And pnpm will print out the warning message when you are not using the corresponding version of node.[^1]

IIRC, Volta does not support reading the `engines` property and pinning node as per its specified version yet. There have been tracking issues about supporting `engines`[^2]. And as one of the users of Volta, I wish it will be eventually supported, too.

For now, to avoid pnpm's warning message, you can pin a node version that is identical to the one of `engines` by `volta pin node@19.x.x`, or delete the `engines` property.

```diff
+  "volta": {
+    "node": "18.13.0",
+    "pnpm": "7.27.1"
+  },
- "engines": {
-   "node": ">=19"
-  }
```

[^1]: https://github.com/pnpm/pnpm/issues/3673
[^2]: https://github.com/volta-cli/volta/issues/1422

**Comment by michaelhays** [comment]
Is there somewhere we're tracking work/requirements on the [known limitations](https://docs.volta.sh/advanced/pnpm) of pnpm support?

I'd love to help on those -- from my understanding, we just have the global installations and automatic migrations from the Volta `pnpm` global left until this can go from experimental to stable.

(FWIW, I've been using `pnpm` with `VOLTA_FEATURE_PNPM` set (see #1394) for a couple of months now and it's been working perfectly)

**Comment by lolmaus** [comment]
I was running into this error:

```
❯ volta install pnpm
error: Executable 'pnpm' is already installed by corepack
```

I resolved it by using this command instead:

```
corepack prepare pnpm@latest --activate
```

**Comment by dwiyatci** [comment]
@chawyehsu Seems like I got into this awkward state in a project where:

```json
"volta": {
  "node": "18.18.2"
},
```

But then when I ran: `pnpm`, it threw: `"Volta error: No pnpm version found in this project."`.

I've got **volta v1.1.1** and `export VOLTA_FEATURE_PNPM=0` set in my `.zshrc`.

I've read this bit from https://docs.volta.sh/advanced/pnpm#migrating, but I don't quite get it:
> Once you switch to the native pnpm support, you may not remove the isolated old pnpm package by calling the same command because lack of uninstall implementation as of now.

Running `which pnpm` outside of the project results in `/Users/Glenn/Library/pnpm/pnpm`.
But running `which pnpm` from the project results in `/Users/Glenn/.volta/bin/pnpm`.

My intention is to keep using globally-installed `pnpm` without `volta` gets in the way. How can I achieve that back? Thank you 🙏

**Comment by chawyehsu** [comment]
> keep using globally-installed pnpm without volta gets in the way.

@dwiyatci There is nothing `VOLTA_FEATURE_PNPM`  can do, Volta can't deal with pnpm that's not installed with Volta. You want to use pnpm outside of Volta then you may need to remove Volta's shims from PATH.

**Comment by dwiyatci** [comment]
> you may need to remove Volta's pnpm shims from PATH.

Thanks for the pointer, @chawyehsu! I did the following and seems like I managed to switch back using global pnpm properly:
- Remove `pnpm` entry from `/Users/Glenn/.volta/tools/user/platform.json`
- Remove `/Users/Glenn/.volta/bin/pnpm`
- (Re-)install `pnpm` (globally): `curl -fsSL https://get.pnpm.io/install.sh | sh -`

Have a nice weekend! :)🎉

**Comment by reowl666** [comment]
It's almost four years since this issue was raised, but why is it still supported as an experiment and not made official?

Please let us know if there are any unresolved issues.
Volta won't support pnpm forever, so I'm starting to avoid it.

**Comment by ShravanSunder** [comment]
i'm also curious about the experimental tag.

**Comment by chawyehsu** [comment]
The project is fully community driven now. It's ok that you going to avoid this avoid that, your business is your business. However that does not help at all just complaining and pressuring in this thread. Read the source code, know how it work and what is unresolved, and submit PRs to contribute, that is the way! And that's how I did and added the existing pnpm support.

**Comment by smblee** [comment]
FYI I use https://mise.jdx.dev/ now instead of volta.

**Comment by BleedingDev** [comment]
And now I switched to Moon. :) Would be good to see some comparison.

https://moonrepo.dev/proto

**Comment by TheBrenny** [comment]
> Read the source code, know how it work and what is unresolved, and submit PRs to contribute, that is the way!

Totally agree! However, there's 4 years of lessons learnt and large amounts of lore in this thread that people jumping in aren't tracking. Yes they can read the thread, but then there's also things stuck in brains that we don't have access to (such as why certain decisions were made).

I don't know Rust, and my patience for it is incredibly low 🙁, but in the interest of the community, are people able to brain-dump what we all know about this issue, it's challenges, opportunities, and the decisions made throughout the dev process? Perhaps @willsoto could then add these points to the original post he made?

> It's ok that you going to avoid this avoid that, your business is your business.

Fwiw, I did move to moonrepo, but it didn't really _feel_ as nice as Volta, and so I've moved back.

**Comment by RoyRao2333** [comment]
> Fwiw, I did move to moonrepo, but it didn't really feel as nice as Volta, and so I've moved back.

They are both great tools. But the reason I chose Proto over Volta is not only because it supports pnpm, but also because of its positive attitude towards fixing issues and the high frequency of updates and patches. It also has an active Discord community and enthusiastic maintainers.

Not gonna make any personal comments about this repo and its maintainers, because it's a great library all the same, and I admire the people who have maintained it for so long.

But for those people who are just reading this thread, I hope you go ahead and try [moonrepo/proto](https://github.com/moonrepo/proto).

**Comment by chriskrycho** [comment]
We definitely don’t take it personally – if another tool works better, by all means go for it! And we’re all *painfully* aware of how little activity the repo has. Moonrepo and proto has actual direct funding to work on it, so more power to them! Volta does *not* have any ongoing funding, and indeed has only had a few months of active funding since it got to the “solves all of the primary corporate sponsor’s internal needs” point back in late 2020 (and those couple months of work were for solving another of those internal needs).

This is, alas, an all-too-familiar tradeoff of corporate-sponsored OSS: it is really hard to justify further work on it when it solves the problems folks have in whatever corporate context. Insofar as you see work happening on Volta *at all*, it’s on our personal time—which, in my case for example, is also competing against family time, [other non-software projects 🎼](https://music.chriskrycho.com), and hobbies besides my job 🏃🏻‍♂️🚴🏼‍♂️!

We are doing what we can even so, but again, we’re *not* offended if people find the velocity and community engagement of other actively funded tools more to their taste: that’s perfectly reasonable and understandable!

**Comment by Szorcu** [comment]
Sharing a solution that, so far, worked really well for me. 
I ended up using **volta** strictly as a node version manager, which I still believe is the best option out there, and switched for **corepack**, the build-in node tool, for managing pnpm. Both tools let you pin a specific node/pnpm version for the project, which is essential for me, especially when working with larger teams.

For setup, there's a great article that covers that in details  (https://www.yanglinzhao.com/til/integrating-corepack-and-volta), but essentially all you need are these few commands:

```sh
volta install corepack
corepack enable --install-directory ~/.volta/bin

volta pin node@24 # pins node version in package.json
corepack use pnpm@10 # pins pnpm version in package.json
```
